### PR TITLE
[mpqsolver] Introduce Evaluator

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit introduces Evaluator for the task of automatic mixed precision quantization.

Draft: https://github.com/Samsung/ONE/pull/10191
Related to: https://github.com/Samsung/ONE/issues/10154

Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    // Get output.
+    const auto output_nodes = loco::output_nodes(trgt_fq_module->graph());
+    for (size_t out_index = 0; out_index < _ref_output[sample_index].size(); ++out_index)
+    {
+      const auto *output_node =
+        loco::must_cast<const luci::CircleOutput *>(output_nodes[out_index]);
+      loco::DataType out_type = output_node->dtype();
+      const ElementaryOutput &ref_elementary = _ref_output[sample_index][out_index];
+      const ElementaryOutput &cur_elementary = cur_output[sample_index][out_index];
+      size_t cur_size = ref_elementary.size() / loco::size(out_type);
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:130]:
```suggestion
      assert(ref_elementary.size() == cur_elementary.size());
      size_t cur_size = ref_elementary.size() / loco::size(out_type);
```

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+          {
+            float ref_value = *(ref_floats + index);
+            float cur_value = *(cur_floats + index);
+            accuracy += std::fabs(ref_value - cur_value);
+          }
+          output_size += cur_size;
+        }
+        break;
+        default:
+          throw std::runtime_error("Unknown out_type");
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:148]:
```suggestion
          throw std::runtime_error("Unsupported output dtype. " + output_node->name());
```

File: compiler/circle-mpqsolver/src/bisection/Evaluator.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <string>
+#include <vector>
+
+namespace mpqsolver
+{
+namespace bisection
+{
+
+using ElementaryOutput = std::vector<char>;
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.h:31]: Better to rename the variable.

```suggestion
using Buffer = std::vector<char>;
```

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.h:31]: Ahh... Ok.

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    interpreter.interpret();
+
+    NNOutput nn_output;
+
+    // Get output.
+    const auto output_nodes = loco::output_nodes(module->graph());
+    for (int i = 0; i < module->graph()->outputs()->size(); i++)
+    {
+      const auto *output_node = loco::must_cast<const luci::CircleOutput *>(output_nodes[i]);
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:94]: Please check output_node's dtype is supported type (float32 as of now).

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp (hunk 2 above)
Comment (jinevening) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp]: Hmm.. It seems that switch statement is unnecessary. Only fp32 dtype can be used in this case. Am I right?

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp (hunk 4 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:94]: ok

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+} // namespace
+
+DatasetEvaluator::DatasetEvaluator(const luci::Module *ref_module, const std::string &h5file)
+  : _ref_module(ref_module), _h5file(h5file)
+{
+  _ref_output = compute_outputs(_ref_module, _h5file);
+}
+
+float DatasetEvaluator::evaluate(const luci::Module *trgt_fq_module) const
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:114]: Currenlty evaluate only returns Mean Absolute Error. Other metrics may be added later.

Could you make the metric configurable?

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp (hunk 2 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp]: Yes. You are right. I'll remove switch.

File: compiler/circle-mpqsolver/src/bisection/Evaluator.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+using NNOutput = std::vector<ElementaryOutput>;
+using DatasetOutput = std::vector<NNOutput>;
+
+class DatasetEvaluator final
+{
+public:
+  DatasetEvaluator(const luci::Module *ref_module, const std::string &h5file);
+  DatasetEvaluator() = delete;
+  ~DatasetEvaluator() = default;
+  float evaluate(const luci::Module *trgt_fq_module) const;
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.h:41]: Please add comments the role of `DatasetEvaluator`, constructor, and `evaluate`.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.h:41]: ok.

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp (hunk 5 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:114]: yes, i can. I'll add parameter to the constructor.

Comment (jinevening) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:114]: One way to support configurable metric is to make a `Metric` class which has interface. `MetricPrinter.h` in circle-eval-diff may be helpful.

File: compiler/circle-mpqsolver/CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +9,8 @@ target_link_libraries(circle-mpqsolver safemain)
 target_link_libraries(circle-mpqsolver luci_lang)
 target_link_libraries(circle-mpqsolver luci_service)
 target_link_libraries(circle-mpqsolver luci_pass)
+target_link_libraries(circle-mpqsolver luci_interpreter)
+target_link_libraries(circle-mpqsolver dio_hdf5)
```
Comment (jinevening) [inline compiler/circle-mpqsolver/CMakeLists.txt:14]: Nor for this PR, but you may add `nncc_common` library and resolve compile warnings.

```suggestion
target_link_libraries(circle-mpqsolver nncc_common)

```

Comment (stamalakhov) [reply compiler/circle-mpqsolver/CMakeLists.txt:14]: ok.

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp (hunk 5 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:114]: > One way to support configurable metric is to make a `Metric` class which has interface. `MetricPrinter.h` in circle-eval-diff may be helpful.

Ahh.. I see. Seems like it's more preferable.

File: compiler/circle-mpqsolver/src/bisection/Evaluator.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    loco::DataType out_dtype = output_node->dtype();
+    if (out_dtype != loco::DataType::FLOAT32)
+      throw std::runtime_error("Unsupported output dtype. " + output_node->name());
+  }
+}
+
+float DatasetEvaluator::evaluate(const luci::Module *trgt_fq_module) const
+{
+  if (_metric == nullptr)
+    throw std::runtime_error("Unsupported metric. ");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:127]: Please remove empty space at the end.
- maybe `Invalid metric` ?

optional) null check for `trgt_fq_module`

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:127]: Ahh... sorry. ok.

Comment (seanshpark) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp:127]: I think `_metric` validation should be done when set, not when used.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/Evaluator.cpp]: fixed

File: compiler/circle-mpqsolver/src/bisection/ErrorMetric.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+using namespace mpqsolver::bisection;
+
+/**
+ * @brief compare first and second operands in MAE (Mean Average Error metric)
+ */
+float MAEMetric::compute(const WholeOutput &first, const WholeOutput &second) const
+{
+  float error = 0.f;
+  size_t output_size = 0;
+
```
Comment (mhs4670go) [inline compiler/circle-mpqsolver/src/bisection/ErrorMetric.cpp:34]: Adding an assert `first.size()` == `second.size()` or throwing an error would be safer.

File: compiler/circle-mpqsolver/src/bisection/ErrorMetric.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  for (size_t sample_index = 0; sample_index < first.size(); ++sample_index)
+  {
+    for (size_t out_index = 0; out_index < first[sample_index].size(); ++out_index)
```
Comment (mhs4670go) [inline compiler/circle-mpqsolver/src/bisection/ErrorMetric.cpp:37]:
```suggestion
  {
    assert (first[sample_index].size() == second[sample_index].size());
    for (size_t out_index = 0; out_index < first[sample_index].size(); ++out_index)
```

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/circle-mpqsolver/src/bisection/Evaluator.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * @brief throws if there is something wrong with the module
+   */
+  void validate(const luci::Module *module) const;
+
+private:
+  const luci::Module *_ref_module = nullptr;
+  std::string _h5file;
+  WholeOutput _ref_output;
+  const ErrorMetric *_metric = nullptr;
```
Comment (jinevening) [inline compiler/circle-mpqsolver/src/bisection/Evaluator.h:60]: (optional) Using `unique_ptr` for `_metric` would be better to manage lifetime of `ErrorMetric` instance.

```suggestion
  std::uniqe_ptr<ErrorMetric> _metric = nullptr;
```

**Review by jinevening — APPROVED** [review]

**Pull request merged**